contrib/pkg/input/input_reconciler.go

```diff
@@ -0,0 +1,135 @@
+	singleClusterReconcileFunc SingleClusterReconcileFunc
+}
+
+// Note(ilackarms): in the current implementation, the constructor
+// also starts the reconciler's event processor in a goroutine.
+// Make sure to cancel the parent context in order to ensure the goroutine started here is gc'ed.
+func NewMultiClusterReconcilerImpl(
+	ctx context.Context,
+	reconcileFunc MultiClusterReconcileFunc,
+) MultiClusterReconciler {
+	r := &inputReconciler{
+		ctx:                       ctx,
+		queue:                     workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		multiClusterReconcileFunc: reconcileFunc,
+	}
+	go r.reconcileEventsForever()
+	return r
+}
+
+// Note(ilackarms): in the current implementation, the constructor
+// also starts the reconciler's event processor in a goroutine.
+// Make sure to cancel the parent context in order to ensure the goroutine started here is gc'ed.
+func NewSingleClusterReconciler(
+	ctx context.Context,
+	reconcileFunc SingleClusterReconcileFunc,
+) SingleClusterReconciler {
+	r := &inputReconciler{
+		ctx:                        ctx,
+		queue:                      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		singleClusterReconcileFunc: reconcileFunc,
+	}
+	go r.reconcileEventsForever()
+	return r
+}
+
+func (r *inputReconciler) ReconcileGeneric(id ezkube.ResourceId) (reconcile.Result, error) {
+	if r.ctx == nil {
+		return reconcile.Result{}, eris.Errorf("internal error: reconciler not started")
+	}
+	contextutils.LoggerFrom(r.ctx).Debugw("reconciling event", "id", sets.Key(id))
+	r.queue.AddRateLimited(id)
+
+	return reconcile.Result{}, nil
+}
+
+func (r *inputReconciler) ReconcileClusterGeneric(id ezkube.ClusterResourceId) (reconcile.Result, error) {
+	if r.ctx == nil {
+		return reconcile.Result{}, eris.Errorf("internal error: reconciler not started")
+	}
+	contextutils.LoggerFrom(r.ctx).Debugw("reconciling event", "cluster", id.GetClusterName(), "id", sets.Key(id))
+	r.queue.AddRateLimited(id)
+
+	return reconcile.Result{}, nil
+}
+
+// reconcile queued events until context is cancelled.
+// blocking (runs from a goroutine)
+func (r *inputReconciler) reconcileEventsForever() {
+	for r.processNextWorkItem() {
+		select {
+		case <-r.ctx.Done():
+			return
+		default:
+		}
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *inputReconciler) processNextWorkItem() bool {
+	key, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(key)
+
+	var err error
+	switch {
+	case r.singleClusterReconcileFunc != nil:
+		err = r.singleClusterReconcileFunc(key.(ezkube.ResourceId))
+	case r.multiClusterReconcileFunc != nil:
+		err = r.multiClusterReconcileFunc(key.(ezkube.ClusterResourceId))
+	}
+
+	if err == nil {
+		r.queue.Forget(key)
+	} else {
+		contextutils.LoggerFrom(r.ctx).Errorw("encountered error reconciling state; retrying", "error", err)
+		r.queue.AddRateLimited(key)
```

Comment:
Can we also allow for a requeue with no error. In `gloo-fed` we have already encountered a situation where returning an error creates an absolute mess in the logs for no reason, because the only return option is an error, and error logging is REALLY verbose.